beacon-chain/rpc/beacon/attestations.go

```diff
@@ -134,44 +147,46 @@ func (bs *Server) ListIndexedAttestations(
 			NextPageToken:       strconv.Itoa(0),
 		}, nil
 	}
-
-	// We use the retrieved committees for the epoch to convert all attestations
+	// We use the retrieved committees for the block root to convert all attestations
 	// into indexed form effectively.
+	mappedAttestations := mapAttestationToBlockRoot(ctx, attsArray)
 	indexedAtts := make([]*ethpb.IndexedAttestation, numAttestations, numAttestations)
-	for i := 0; i < len(atts); i++ {
-		att := atts[i]
-		epoch := helpers.SlotToEpoch(att.Data.Slot)
-		attState, err := bs.StateGen.StateByRoot(ctx, bytesutil.ToBytes32(att.Data.BeaconBlockRoot))
-		if err != nil {
-			return nil, status.Errorf(
-				codes.Internal,
-				"Could not retrieve state for attestation data block root %v: %v",
-				att.Data.BeaconBlockRoot,
-				err,
-			)
-		}
-		activeIndices, err := helpers.ActiveValidatorIndices(attState, epoch)
-		if err != nil {
-			return nil, status.Errorf(
-				codes.Internal,
-				"Could not retrieve active validator indices for epoch %d: %v",
-				epoch,
-				err,
-			)
+	for br, atts := range mappedAttestations {
+		var attState *stateTrie.BeaconState
+		if !featureconfig.Get().DisableNewStateMgmt {
+			attState, err = bs.StateGen.StateByRoot(ctx, br)
+			if err != nil {
+				return nil, status.Errorf(
+					codes.Internal,
+					"Could not retrieve state for attestation data block root %v: %v",
+					br,
+					err,
+				)
+			}
+		} else {
+			attState, err = bs.BeaconDB.State(ctx, br)
+			if err != nil {
```

Comment:
This error is duplicated it seems, maybe you can move it outside of the if statement?